api/include/opentelemetry/common/spin_lock_mutex.h

```diff
@@ -32,22 +55,42 @@ class SpinLockMutex
    */
   void lock() noexcept
   {
-    /* Note: We expect code protected by this lock to be "fast", i.e. we do NOT incrementally
-     * back-off and wait/notify here, we just loop until we have access, then try again.
-     *
-     * This has the downside that we could be spinning a long time if the exporter is slow.
-     * Note: in C++20x we could use `.wait` to make this slightly better. This should move to
-     * an exponential-back-off / wait strategy.
-     */
-    while (flag_.test_and_set(std::memory_order_acquire))
-      /** TODO - We should immmediately yield if the machine is single processor. */
-      ;
+    for (;;)
+    {
+      // Try once
+      if (!flag_.exchange(true, std::memory_order_acquire))
+      {
+        return;
+      }
+      // Spin-Fast
+      for (std::size_t i = 0; i < SPINLOCK_FAST_ITERATIONS; ++i)
+      {
+        if (try_lock())
+        {
+          return;
+        }
+        // TODO: Issue PAUSE/YIELD instruction to reduce contention.
+        // e.g. __builtin_ia32_pause() / YieldProcessor() / _mm_pause();
+      }
+      // Spin-Yield
+      for (std::size_t i = 0; i < SPINLOCK_YIELD_ITERATIONS; ++i)
+      {
+        if (try_lock())
+        {
+          return;
+        }
+        std::this_thread::yield();
+      }
+      // Sleep and then start the whole process again.
```

Comment:
As there is an unconditional loop anyway, wondering why not do that in the last Spin-Yield instead of retrying Spin-Fast? Is there some benefit to do Spin-Fast after Spin-Yield timeout?

Reply:
The idea is that (theoretically) spin fast is on the order of 10s of ns, while yield is 100s of NS and the full sleep is 1ms, i.e. each is an order of magnitude slower.  The goal is to "catch" the lock when you're scheduled and no need to give up to another thread if you only need to wait a few ns, not a big deal.

This theory could be wrong and we should validate with a benchmark the behavior we want.